packages/babel-parser/src/plugins/flow/index.js

```diff
@@ -2219,6 +2219,22 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       }
     }
 
+    isIterator(word: string): boolean {
+      return word === "iterator" || word === "asyncIterator";
+    }
+
+    readIterator(): void {
+      const word = super.readWord1();
+      const fullWord = "@@" + word;
+
+      // Allow @@iterator and @@asyncIterator as a identifier only inside type
+      if (!this.isIterator(word) || !this.state.inType) {
+        this.raise(this.state.pos, Errors.InvalidIdentifier, fullWord);
+      }
```

Comment:
The implementation of `readIterator` is copied from original `readWord`. It oversights that an iterator identifier may contain escapes and thus should not be allowed:

For example, currently Babel parses successfully, which should have thrown.
```js
function foo(): { @@iter\u0061tor: () => string } {
  return (0: any);
}
```
This PR focuses on performance only and I will open a new PR for that after this PR gets merged.